File:  貪心思想/1833. Maximum Ice Cream Bars.java
import java.util.Arrays;

/**
 * 吃冰的問題
 * 
 * Input: costs = [1,3,2,4,1], coins = 7
 * Output: 4
 * Explanation: The boy can buy ice cream bars at indices 0,1,2,4 for a total
 * price of 1 + 3 + 2 + 1 = 7.
 * 
 * Input: costs = [10,6,8,7,7,8], coins = 5
 * Output: 0
 * Explanation: The boy cannot afford any of the ice cream bars.
 * 
 * Input: costs = [1,6,3,1,2,5], coins = 20
 * Output: 6
 * Explanation: The boy can buy all the ice cream bars for a total price of 1 +
 * 6 + 3 + 1 + 2 + 5 = 18.
 */
class Solution {
    public int maxIceCream(int[] costs, int coins) {
        if (costs.length == 0)
            return 0;
        Arrays.sort(costs);

        int ans = 0;
        for (int i = 0; i < costs.length && coins >= 0; i++) {
            coins -= costs[i];
            ans++;
        }

        if (coins < 0) {
            ans--;
        }

        return ans;
    }
}

File:  貪心思想/1833. Maximum Ice Cream Bars.kt

"1833. Maximum Ice Cream Bars" 是一个贪心算法问题。问题的描述如下：

你有一笔钱 coins 和一个冰淇淋数组 costs，其中 costs[i] 表示第 i 种冰淇淋的价格。你可以购买任意数量的这些冰淇淋，
但必须在 coins 内的预算范围内。要最大化你能购买的冰淇淋数量。

要解决这个问题，你可以按照以下步骤使用贪心算法：

1. 将 costs 数组按照价格从低到高进行排序。
2. 从最便宜的冰淇淋开始购买，直到你的预算 coins 不足以购买下一种冰淇淋为止。
以下是使用 Kotlin 解决这个问题的示例代码：

class Solution {
    fun maxIceCream(costs: IntArray, coins: Int): Int {
        costs.sort() // 按价格排序
        var count = 0 // 初始化购买的冰淇淋数量
        var remainingCoins = coins // 初始化剩余预算
        
        for (cost in costs) {
            if (cost <= remainingCoins) {
                // 如果当前冰淇淋的价格在预算范围内
                count++
                remainingCoins -= cost
            } else {
                // 如果预算不足以购买当前冰淇淋，停止购买
                break
            }
        }
        
        return count
    }
}

这个代码首先对冰淇淋价格数组 costs 进行排序，以便从最便宜的冰淇淋开始购买。

然后，我们初始化购买的冰淇淋数量 count 和剩余预算 remainingCoins，开始遍历排序后的冰淇淋价格。

在遍历过程中，如果当前冰淇淋的价格 cost 小于或等于剩余预算 remainingCoins，我们将购买该冰淇淋，增加 count 计数器，并减少剩余预算。

如果当前冰淇淋的价格已经超出了剩余预算，我们停止购买，因为无法购买更贵的冰淇淋。

最后，返回购买的冰淇淋数量 count，它表示在给定预算下最多可以购买的冰淇淋数量。

这个算法的时间复杂度是 O(nlogn)，其中 n 是冰淇淋的种类数量，因为我们需要对价格数组进行排序。希望这个解释和示例代码有助于理解 "1833. Maximum Ice Cream Bars" 问题的解决方法。